tests/kafkatest/tests/streams/streams_multiple_rolling_upgrade_test.py

```diff
@@ -0,0 +1,110 @@
+                                                                    'dif': {'partitions': 5, 'replication-factor': 1},
+                                                                    'cnt': {'partitions': 5, 'replication-factor': 1},
+                                                                    'avg': {'partitions': 5, 'replication-factor': 1},
+                                                                    'wcnt': {'partitions': 5, 'replication-factor': 1},
+                                                                    'tagg': {'partitions': 5, 'replication-factor': 1}
+                                                                })
+
+        self.driver = StreamsSmokeTestDriverService(test_context, self.kafka)
+        self.processor_1 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
+        self.processor_2 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
+        self.processor_3 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
+
+        # already on trunk version at end of upgrades so get rid of it
+        self.streams_downgrade_versions = self.streams_upgrade_versions[:-1]
+        self.streams_downgrade_versions.reverse()
+
+        self.processors = [self.processor_1, self.processor_2, self.processor_3]
+
+        self.started = False
+
+    def setUp(self):
+        self.zk.start()
+
+    def upgrade_and_verify_start(self, processors, to_version):
+        for processor in processors:
+            self.logger.info("Updating node %s to version %s" % (processor.node.account, to_version))
+            node = processor.node
+            if self.started:
+                processor.stop()
+                self.wait_for_verification(processor, "SMOKE-TEST-CLIENT-CLOSED", processor.STDOUT_FILE)
+            node.version = KafkaVersion(to_version)
+            processor.start()
+            self.wait_for_verification(processor, "initializing processor: topic", processor.STDOUT_FILE)
+
+        self.started = True
+
+    def update_processors_and_verify(self, versions):
+        for version in versions:
+            self.upgrade_and_verify_start(self.processors, version)
+        self.run_data_and_verify()
+
+    def run_data_and_verify(self):
+        self.driver.start()
+        self.wait_for_verification(self.driver, "ALL-RECORDS-DELIVERED", self.driver.STDOUT_FILE)
+        self.driver.stop()
+
+    @cluster(num_nodes=9)
+    @matrix(broker_version=streams_upgrade_versions)
+    def test_rolling_upgrade_downgrade_multiple_apps(self, broker_version):
+        self.kafka.set_version(KafkaVersion(broker_version))
+        self.kafka.start()
+
+        # verification step run after each upgrade
+        self.update_processors_and_verify(self.streams_upgrade_versions)
+
+        # with order reversed now we test downgrading, verification run after each downgrade
+        self.update_processors_and_verify(self.streams_downgrade_versions)
```

Comment:
Don't we need to stop all processor at the end?

Reply:
ack